### ISSUE TITLE
CmaEsSampler uses independent sampler even for non-dynamic search spaces

### ISSUE THREAD

**Issue by Callidior** [optuna/optuna#3316]
### Expected behavior

I am using a clearly non-dynamic search space with the `CmaEsSampler` but still see warnings that the independent sampler is triggered very frequently.

I investigated this with a debugger and found that `CmaEsSampler` sometimes (or in my case, frequently) samples values that are just a bit outside of the distribution bounds. `Trial._is_relative_param()` then checks whether the value sampled for the relative search space is within bounds and, if not, the independent sampler is triggered.

#1557 seems to be a related issue.

My `Trial.relative_search_space` looks as follows:

```python
{
    'initial_learning_rate': LogUniformDistribution(high=0.001, low=1e-05),
    'learning_rate_decay_factor': UniformDistribution(high=0.9999, low=0.8),
    'base_filter_size_ln2': IntUniformDistribution(high=6, low=3, step=1),
    'block_size': IntUniformDistribution(high=2, low=1, step=1),
    'bottleneck_size': IntUniformDistribution(high=150, low=10, step=10),
    'num_blocks': IntUniformDistribution(high=5, low=2, step=1),
    'num_downsampling_steps': IntUniformDistribution(high=7, low=3, step=1),
    'reg_scale_l1': LogUniformDistribution(high=0.1, low=0.0001),
    'number_of_epochs': IntUniformDistribution(high=1000, low=500, step=100),
    'weight_decay': LogUniformDistribution(high=0.001, low=1e-06)
}
```

And the trial I investigated with the debugger had the following `Trial.relative_params` sampled with CmaEsSampler:

```python
{
    'initial_learning_rate': 9.999999999999997e-06,
    'learning_rate_decay_factor': 0.9998999999999999,
    'base_filter_size_ln2': 7,
    'block_size': 1,
    'bottleneck_size': 150,
    'num_blocks': 2,
    'num_downsampling_steps': 3,
    'reg_scale_l1': 0.00010000000000000009,
    'number_of_epochs': 500,
    'weight_decay': 0.0009999999999999998
}
```

We can see that some sampled values are slightly out of bounds:

- `initial_learning_rate` is 9.999e-06 < 1e-05.
- `base_filter_size_ln2` is 7 > 6.

This happens on almost every trial for me. Especially the integer parameters are prone to this.
Is this a bug or do I need to make my search space less restricted?

### Environment

- Optuna version: 2.10.0
- Python version: 3.8.12
- OS: Linux-5.4.0-94-generic-x86_64-with-glibc2.10
- (Optional) Other libraries and their versions:
  - cmaes: 0.8.2

### Error messages, stack traces, or logs

```shell
The parameter 'number_of_epochs' in trial#217 is sampled independently by using `TPESampler` instead of `CmaEsSampler` (optimization performance may be degraded). `CmaEsSampler` does not support dynamic search space or `CategoricalDistribution`. You can suppress this warning by setting `warn_independent_sampling` to `False` in the constructor of `CmaEsSampler`, if this independent sampling is intended behavior.
The parameter 'base_filter_size_ln2' in trial#217 is sampled independently by using `TPESampler` instead of `CmaEsSampler` (optimization performance may be degraded). `CmaEsSampler` does not support dynamic search space or `CategoricalDistribution`. You can suppress this warning by setting `warn_independent_sampling` to `False` in the constructor of `CmaEsSampler`, if this independent sampling is intended behavior.
The parameter 'num_downsampling_steps' in trial#217 is sampled independently by using `TPESampler` instead of `CmaEsSampler` (optimization performance may be degraded). `CmaEsSampler` does not support dynamic search space or `CategoricalDistribution`. You can suppress this warning by setting `warn_independent_sampling` to `False` in the constructor of `CmaEsSampler`, if this independent sampling is intended behavior.
The parameter 'num_blocks' in trial#217 is sampled independently by using `TPESampler` instead of `CmaEsSampler` (optimization performance may be degraded). `CmaEsSampler` does not support dynamic search space or `CategoricalDistribution`. You can suppress this warning by setting `warn_independent_sampling` to `False` in the constructor of `CmaEsSampler`, if this independent sampling is intended behavior.
The parameter 'block_size' in trial#217 is sampled independently by using `TPESampler` instead of `CmaEsSampler` (optimization performance may be degraded). `CmaEsSampler` does not support dynamic search space or `CategoricalDistribution`. You can suppress this warning by setting `warn_independent_sampling` to `False` in the constructor of `CmaEsSampler`, if this independent sampling is intended behavior.
```

### Steps to reproduce

Admittedly, the following MWE is not optimal, but I could reproduce the behavior with it at least one out of 100 times:

```python
import optuna

class MWE:
    def __call__(self, trial):
        num = trial.suggest_int("num", 0, 15, step=1)
        denom = trial.suggest_int("denom", 1, 2, step=1)
        x = num / denom
        return 6 * (x**4) - 14 * (x**3) - 2 * (x**2) + 10 * x

if __name__ == "__main__":
    optuna.logging.set_verbosity(optuna.logging.WARNING)
    objective = MWE()

    for seed in range(100):
        study = optuna.create_study(
            sampler=optuna.samplers.CmaEsSampler(seed=seed)
        )
        study.optimize(objective, n_trials=500)
        print("Best value: {} (params: {})\n".format(study.best_value, study.best_params))
```

This yielded:

```
[W 2022-02-15 13:36:30,988] The parameter 'denom' in trial#37 is sampled independently by using `RandomSampler` instead of `CmaEsSampler` (optimization performance may be degraded). `CmaEsSampler` does not support dynamic search space or `CategoricalDistribution`. You can suppress this warning by setting `warn_independent_sampling` to `False` in the constructor of `CmaEsSampler`, if this independent sampling is intended behavior.
```

**Comment by nzw0301** [comment]
Thank you for reporting this unexpected behaviour and sharing your investigation.

Let me put a shorter script to reproduce the results below:

```python
import optuna


def objective(trial):
    num = trial.suggest_int("num", 0, 15, step=1)
    denom = trial.suggest_int("denom", 1, 2, step=1)
    x = num / denom
    return 6 * (x**4) - 14 * (x**3) - 2 * (x**2) + 10 * x

optuna.logging.set_verbosity(optuna.logging.WARNING)

seed = 28

study = optuna.create_study(
    sampler=optuna.samplers.CmaEsSampler(seed=seed)
)
study.optimize(objective, n_trials=40)
print("seed{}, Best value: {} (params: {})\n".format(seed, study.best_value, study.best_params))
```


- Optuna: 3.0.0b0.dev
- Python version: 3.9.7
- OS: macOS-10.16-x86_64-i386-64bit

**Comment by nzw0301** [comment]
I suppose a similar logic to 
https://github.com/optuna/optuna/blob/e71e0f468b33c65cc560ed978b394fceccafacda/optuna/_transform.py#L322
resolve this issue.

**Comment by Callidior** [comment]
Thanks for the quick fix. This should indeed to the job.

**Comment by Callidior** [comment]
I noticed that in my actual experiments (not the MWE), CMA-ES samples extreme values most of the time even for continuous parameters. For instance, learning rate was chosen very close to one of the bounds in 85% of trials.
I would understand this if it was always the same bound but it is evenly distributed across high and low bound.

Is this behavior known or expected? Are 10 parameters too many to optimize jointly with CMA-ES?
Any advice would be highly appreciated :)

**Comment by juroberttyb** [comment]
Me also have the same problem here, using CmaEs for with non-dynamic search space and have been doing well for approximately a hundred trails, then somehow it just start using random sampler for one of the parameter.
